database/src/ops/block.rs

```diff
@@ -1,89 +1,411 @@
+        BlockBlobs, BlockHeights, BlockInfos, KeyImages, NumOutputs, Outputs, PrunableHashes,
+        PrunableTxBlobs, PrunedTxBlobs, RctOutputs, Tables, TablesMut, TxHeights, TxIds,
+        TxUnlockTime,
+    },
+    transaction::{TxRo, TxRw},
+    types::{
+        AmountIndex, BlockHash, BlockHeight, BlockInfo, KeyImage, Output, OutputFlags,
+        PreRctOutputId, RctOutput, TxHash,
+    },
+    StorableVec,
+};
 
-/// TODO
-pub fn block_exists() {
-    todo!()
-}
+//---------------------------------------------------------------------------------------------------- `add_block_*`
+/// Add a [`VerifiedBlockInformation`] to the database.
+///
+/// This extracts all the data from the input block and
+/// maps/adds them to the appropriate database tables.
+///
+#[doc = doc_error!()]
+///
+/// # Panics
+/// This function will panic if:
+/// - `block.height > u32::MAX` (not normally possible)
+/// - `block.height` is not != [`chain_height`]
+///
+/// # Already exists
+/// This function will operate normally even if `block` already
+/// exists, i.e., this function will not return `Err` even if you
+/// call this function infinitely with the same block.
+// no inline, too big.
+pub fn add_block(
+    block: &VerifiedBlockInformation,
+    tables: &mut impl TablesMut,
+) -> Result<(), RuntimeError> {
+    //------------------------------------------------------ Check preconditions first
 
-/// TODO
-pub fn get_block_hash() {
-    todo!()
-}
+    // Cast height to `u32` for storage (handled at top of function).
+    // Panic (should never happen) instead of allowing DB corruption.
+    // <https://github.com/Cuprate/cuprate/pull/102#discussion_r1560020991>
+    let Ok(height) = u32::try_from(block.height) else {
+        panic!("block.height ({}) > u32::MAX", block.height);
+    };
 
-/// TODO
-pub fn get_block_height() {
-    todo!()
-}
+    let chain_height = chain_height(tables.block_heights())?;
+    assert_eq!(
+        block.height, chain_height,
+        "block.height ({}) != chain_height ({})",
+        block.height, chain_height,
+    );
 
-/// TODO
-pub fn get_block_weight() {
-    todo!()
-}
+    //------------------------------------------------------ Transaction / Outputs / Key Images
+    for tx_verification_data in &block.txs {
+        add_tx(tx_verification_data, &chain_height, tables)?;
+    }
 
-/// TODO
-pub fn get_block_already_generated_coins() {
-    todo!()
-}
+    //------------------------------------------------------ Block Info
 
-/// TODO
-pub fn get_block_long_term_weight() {
-    todo!()
-}
+    // INVARIANT: must be below the above transaction loop since this
+    // RCT output count needs account for _this_ block's outputs.
+    let cumulative_rct_outs = get_rct_num_outputs(tables.rct_outputs())?;
+
+    // Block Info.
+    tables.block_infos_mut().put(
+        &block.height,
+        &BlockInfo {
+            timestamp: block.block.header.timestamp,
+            total_generated_coins: block.generated_coins,
```

Comment:
`total_generated_coins` is a cumulative value btw, so this should be adding the previous blocks total_generated_coins

Reply:
could maybe rename to make it more clear? up to you